cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -756,211 +1078,227 @@ class StructConverter : public TypedConverterVisitor<StructBuilder, StructConver
   // Whether we're converting from a sequence of dicts or tuples
   enum { UNKNOWN, DICTS, TUPLES } source_kind_ = UNKNOWN;
   bool from_pandas_;
+  bool strict_conversions_;
 };
 
-class DecimalConverter
-    : public TypedConverterVisitor<arrow::Decimal128Builder, DecimalConverter> {
+class DecimalConverter : public TypedConverter<arrow::Decimal128Type, DecimalConverter> {
  public:
-  // Append a non-missing item
+  using BASE = TypedConverter<arrow::Decimal128Type, DecimalConverter>;
+  Status Init(ArrayBuilder* builder) override {
+    RETURN_NOT_OK(BASE::Init(builder));
+    decimal_type_ = checked_cast<const DecimalType*>(typed_builder_->type().get());
+    return internal::ImportDecimalType(&pydecimal_type_);
+  }
+
   Status AppendItem(PyObject* obj) {
-    Decimal128 value;
-    const auto& type = checked_cast<const DecimalType&>(*typed_builder_->type());
-    RETURN_NOT_OK(internal::DecimalFromPythonDecimal(obj, type, &value));
-    return typed_builder_->Append(value);
+    if (PyObject_IsInstance(obj, pydecimal_type_.obj()) == 1) {
```

Comment:
Is there a reason you don't let `DecimalFromPythonDecimal` do the type check? The additional `PyObject_IsInstance` call will probably slow things down.

(nit: I would write `> 0`)

Reply:
Changing this to use `internal::PyDecimal_Check`